File: android/src/main/kotlin/com/mytiki/sdk/capacitor/Trail.kt
package com.mytiki.sdk.capacitor

import com.getcapacitor.JSObject
import com.getcapacitor.PluginCall
import com.mytiki.tiki_sdk_android.channel.Channel
import com.mytiki.tiki_sdk_android.trail.TrailMethod
import com.mytiki.tiki_sdk_android.trail.Usecase
import com.mytiki.tiki_sdk_android.trail.req.ReqGuard
import com.mytiki.tiki_sdk_android.trail.rsp.RspGuard
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.async

class Trail(private val channel: Channel) {
    val title: Title = Title(channel)
    val license: License = License(channel)
    val payable: Payable = Payable(channel)
    val receipt: Receipt = Receipt(channel)

    fun guard(call: PluginCall) {
        val deferred = MainScope().async {
            val usecases: List<Usecase> =
                call.getArray("usecases").toList<String>().map { usecase -> Usecase.from(usecase) }
            val destinations: List<String> = call.getArray("destinations").toList()
            val rsp: RspGuard = channel.request(
                TrailMethod.GUARD,
                ReqGuard(call.getString("ptr")!!, usecases, destinations)
            ) { args ->
                RspGuard.from(args)
            }.await()
            val ret = JSObject()
            ret.put("success", rsp.success)
            ret.put("reason", rsp.reason)
            call.resolve(ret)
        }
    }
}

File: ios/Plugin/Trail.swift
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

import Foundation
import Capacitor
import TikiSdk

@objc public class Trail: NSObject {
    public var title: Title
    public var license: License
    public var payable: Payable
    public var receipt: Receipt
    
    private var tiki: TikiSdk
    
    init(tiki: TikiSdk) {
        self.tiki = tiki
        self.title = Title(tiki: tiki)
        self.license = License(tiki: tiki)
        self.payable = Payable(tiki: tiki)
        self.receipt = Receipt(tiki: tiki)
    }
    
    @objc public func `guard`(call: CAPPluginCall) async{
        let usecasesArray = call.getArray("usecases") as? [String] ?? []
        let usecases: [Usecase] = usecasesArray.compactMap { usecaseString in
            return Usecase.from(usecase: usecaseString)
        }
        let destinations = call.getArray("destinations") as? [String] ?? []
        let ptr = call.getString("ptr") ?? ""
        do{
            let _ = try await tiki.trail.guard(
                ptr: ptr, usecases: usecases, destinations: destinations,
                onPass: { call.resolve([:]) },
                onFail: { reason in call.reject(reason ?? "Not allowed.") }
                )
        }catch{
            call.reject(error.localizedDescription)
        }
    }
}
